tracing-mock/src/subscriber.rs

```diff
@@ -937,13 +962,32 @@ where
         let was_expected = matches!(expected.front(), Some(Expect::NewSpan(_)));
         if was_expected {
             if let Expect::NewSpan(mut expected) = expected.pop_front().unwrap() {
-                let get_parent_name = || {
-                    span.parent()
-                        .and_then(|id| cx.span(id))
-                        .or_else(|| cx.lookup_current())
-                        .map(|span| span.name().to_string())
+                let get_parent = || {
+                    if span.is_contextual() {
+                        if let Some(parent) = cx.lookup_current() {
+                            let contextual_parent = cx.span(&parent.id()).expect(
+                                "tracing-mock: contextual parent cannot \
+                                be looked up by ID. Was it recorded correctly?",
+                            );
+                            Parent::Contextual(contextual_parent.name().to_string())
+                        } else {
+                            Parent::ContextualRoot
+                        }
+                    } else if span.is_root() {
+                        Parent::ExplicitRoot
+                    } else {
+                        let parent_id = span.parent().expect(
+                            "tracing-mock: is_contextual=false is_root=false \
+                            but no explicit parent found. This is a bug!",
+                        );
+                        let explicit_parent = cx.span(parent_id).expect(
+                            "tracing-mock: explicit parent cannot be looked \
+                            up by ID. Is the provided Span ID valid: {parent_id}?",
+                        );
+                        Parent::Explicit(explicit_parent.name().to_string())
+                    }
```

Comment:
this code is kinda duplicated here and in `collector.rs`, although not _quite_...i wonder if some of it can be factored out? idk. not a blocker.

Reply:
I tried to refactor this a couple of times and didn't come up with anything that wasn't more verbose than this. On top of that, the indirection required to handle different span lookups and context fetching tends to make it all longer as well.

I'd leave tis as is for now if that's OK.